File: roboplan_ros_visualization/bindings/test/roboplan_ik_marker_test.py
#!/usr/bin/env python3

import numpy as np
from visualization_msgs.msg import InteractiveMarkerFeedback

from roboplan.core import Scene, CartesianConfiguration, JointConfiguration
from roboplan.example_models import get_package_models_dir, get_package_share_dir
from roboplan.simple_ik import SimpleIk, SimpleIkOptions
from roboplan.optimal_ik import (
    FrameTask,
    FrameTaskOptions,
    Oink,
    PositionLimit,
)
from roboplan_ros.cpp import se3ToPose
from roboplan_ros.visualization import RoboplanIKMarker


def test_process_feedback():
    models_dir = get_package_models_dir()
    urdf_path = models_dir / "ur_robot_model" / "ur5_gripper.urdf"
    srdf_path = models_dir / "ur_robot_model" / "ur5_gripper.srdf"
    package_paths = [get_package_share_dir()]

    # Just start right next to IK pose to make the IK trivial
    scene = Scene("test", urdf_path, srdf_path, package_paths)
    q_init = np.array(scene.getCurrentJointPositions())
    q_init[:] = [0.0, -1.57, 1.57, -1.57, -1.57, 0.0]
    scene.setJointPositions(q_init)

    # Build a SimpleIK solve function for use in the marker
    joint_group = "arm"
    base_link = "base_link"
    tip_link = "tool0"
    ik_solver = SimpleIk(scene, SimpleIkOptions())
    q_indices = scene.getJointGroupInfo(joint_group).q_indices

    def ik_solve_fn(target_pose, seed):
        goal = CartesianConfiguration()
        goal.base_frame = base_link
        goal.tip_frame = tip_link
        goal.tform = target_pose
        seed_jc = JointConfiguration()
        seed_jc.positions = seed[q_indices]
        solution = JointConfiguration()
        if ik_solver.solveIk(goal, seed_jc, solution):
            return scene.toFullJointPositions(joint_group, solution.positions)
        return None

    ik = RoboplanIKMarker(
        scene=scene,
        base_link=base_link,
        tip_link=tip_link,
        ik_solve_fn=ik_solve_fn,
    )

    # Compute a reachable target via FK at a known configuration
    q_target = np.array(scene.getCurrentJointPositions())
    q_target[q_indices] = [0.0, -1.57, 1.57, -1.57, -1.57, 0.0]
    fk_target = scene.forwardKinematics(q_target, "tool0", "base_link")
    target_pose = se3ToPose(fk_target)

    # Solve via process_feedback
    feedback = InteractiveMarkerFeedback()
    feedback.event_type = InteractiveMarkerFeedback.POSE_UPDATE
    feedback.pose = target_pose
    result = ik.process_feedback(feedback)

    # Verify FK of solution matches the target
    assert result is not None
    fk_result = scene.forwardKinematics(result, "tool0", "base_link")
    assert abs(fk_result[0, 3] - target_pose.position.x) < 0.01
    assert abs(fk_result[1, 3] - target_pose.position.y) < 0.01
    assert abs(fk_result[2, 3] - target_pose.position.z) < 0.01


def test_process_feedback_oink():
    models_dir = get_package_models_dir()
    urdf_path = models_dir / "ur_robot_model" / "ur5_gripper.urdf"
    srdf_path = models_dir / "ur_robot_model" / "ur5_gripper.srdf"
    package_paths = [get_package_share_dir()]

    scene = Scene("test", urdf_path, srdf_path, package_paths)
    q_init = np.array(scene.getCurrentJointPositions())
    q_init[:] = [0.0, -1.57, 1.57, -1.57, -1.57, 0.0]
    scene.setJointPositions(q_init)

    joint_group = "arm"
    base_link = "base_link"
    tip_link = "tool0"

    # Configure OinK solver
    oink = Oink(scene, joint_group)
    num_variables = len(oink.v_indices)
    goal = CartesianConfiguration()
    goal.base_frame = base_link
    goal.tip_frame = tip_link
    task_options = FrameTaskOptions(
        position_cost=1.0,
        orientation_cost=0.1,
        task_gain=1.0,
        lm_damping=0.01,
    )
    frame_task = FrameTask(oink, scene, goal, task_options)
    tasks = [frame_task]

    position_limit = PositionLimit(oink, gain=1.0)
    constraints = [position_limit]
    max_iters = 50
    regularization = 1e-6

    # Setup solve function for IK marker
    def ik_solve_fn(target_pose, seed):
        scene.setJointPositions(seed)
        frame_task.setTargetFrameTransform(target_pose)

        q_current = np.array(seed)
        delta_q = np.zeros(num_variables)
        delta_q_full = np.zeros(len(seed))

        for _ in range(max_iters):
            scene.setJointPositions(q_current)
            scene.forwardKinematics(q_current, tip_link)

            try:
                oink.solveIk(scene, tasks, constraints, [], delta_q, regularization)
            except RuntimeError:
                return None

            if np.linalg.norm(delta_q) < 1e-6:
                break

            delta_q_full[oink.v_indices] = delta_q
            q_current = scene.integrate(q_current, delta_q_full)

        return q_current

    ik = RoboplanIKMarker(
        scene=scene,
        base_link=base_link,
        tip_link=tip_link,
        ik_solve_fn=ik_solve_fn,
    )

    # Compute a reachable target via FK at a known configuration
    q_target = np.array(scene.getCurrentJointPositions())
    q_indices = scene.getJointGroupInfo(joint_group).q_indices
    q_target[q_indices] = [0.0, -1.57, 1.57, -1.57, -1.57, 0.0]
    fk_target = scene.forwardKinematics(q_target, tip_link, base_link)
    target_pose = se3ToPose(fk_target)

    # Solve via process_feedback
    feedback = InteractiveMarkerFeedback()
    feedback.event_type = InteractiveMarkerFeedback.POSE_UPDATE
    feedback.pose = target_pose
    result = ik.process_feedback(feedback)

    # Verify FK of solution matches the target
    assert result is not None
    fk_result = scene.forwardKinematics(result, tip_link, base_link)
    assert abs(fk_result[0, 3] - target_pose.position.x) < 0.01
    assert abs(fk_result[1, 3] - target_pose.position.y) < 0.01
    assert abs(fk_result[2, 3] - target_pose.position.z) < 0.01

File: roboplan_ros_visualization/test/roboplan_ik_marker_test.cpp
#include <gtest/gtest.h>

#include <Eigen/Core>

#include <visualization_msgs/msg/interactive_marker_feedback.hpp>

#include <roboplan/core/scene.hpp>
#include <roboplan_ros_cpp/type_conversions.hpp>
#include <roboplan_ros_visualization/roboplan_ik_marker.hpp>
#include <roboplan_simple_ik/simple_ik.hpp>

namespace roboplan_ros_visualization {

static const std::string TWO_LINK_URDF = R"(<?xml version="1.0"?>
<robot name="test_robot">
  <link name="world"/>
  <link name="link1">
    <visual>
      <geometry>
        <cylinder radius="0.05" length="0.5"/>
      </geometry>
    </visual>
  </link>
  <link name="link2">
    <visual>
      <geometry>
        <cylinder radius="0.05" length="0.5"/>
      </geometry>
    </visual>
  </link>
  <joint name="joint1" type="revolute">
    <parent link="world"/>
    <child link="link1"/>
    <origin xyz="0 0 0.25" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="-3.14" upper="3.14" effort="10" velocity="1"/>
  </joint>
  <joint name="joint2" type="revolute">
    <parent link="link1"/>
    <child link="link2"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="-3.14" upper="3.14" effort="10" velocity="1"/>
  </joint>
</robot>
)";

static const std::string TWO_LINK_SRDF = R"(<?xml version="1.0"?>
<robot name="test_robot">
  <group name="arm">
    <chain base_link="world" tip_link="link2"/>
  </group>
</robot>
)";

class RoboplanIkMarkerTest : public ::testing::Test {};

TEST_F(RoboplanIkMarkerTest, MakeInteractiveMarkerDefaultFrame) {
  auto scene_ = std::make_shared<roboplan::Scene>("test", TWO_LINK_URDF, TWO_LINK_SRDF);

  // Specify links even though it's a silly robot
  const std::string joint_group = "arm";
  const std::string base_link = "world";
  const std::string tip_link = "link2";

  const auto joint_group_info = scene_->getJointGroupInfo(joint_group);
  ASSERT_TRUE(joint_group_info.has_value());
  const auto q_indices = joint_group_info->q_indices;

  // This 2-DOF arm needs a lot of iterating...
  roboplan::SimpleIkOptions ik_options;
  ik_options.max_iters = 1000;

  // Setup IK solver and a solve function to pass to the iMarker
  auto simple_ik = std::make_shared<roboplan::SimpleIk>(
      std::const_pointer_cast<roboplan::Scene>(scene_), ik_options);
  auto ik_solve_fn = [scene_, simple_ik, q_indices, joint_group, base_link,
                      tip_link](const Eigen::Matrix4d& target_pose,
                                const Eigen::VectorXd& seed) -> std::optional<Eigen::VectorXd> {
    roboplan::CartesianConfiguration goal;
    goal.base_frame = base_link;
    goal.tip_frame = tip_link;
    goal.tform = target_pose;

    roboplan::JointConfiguration seed_jc;
    seed_jc.positions = seed(q_indices);

    roboplan::JointConfiguration solution;
    if (simple_ik->solveIk(goal, seed_jc, solution)) {
      return scene_->toFullJointPositions(joint_group, solution.positions);
    }
    return std::nullopt;
  };

  auto ik_ = std::make_unique<RoboplanIKMarker>(scene_, base_link, tip_link, ik_solve_fn);

  // Verify marker construction with default frame
  const auto marker = ik_->construct_imarker();
  EXPECT_EQ(marker.header.frame_id, "world");
  EXPECT_EQ(marker.name, "ik_target");
  EXPECT_FLOAT_EQ(marker.scale, 0.2f);
  EXPECT_EQ(marker.controls.size(), 7u);

  // Compute a reachable target by forward kinematics at a known configuration
  Eigen::VectorXd q_target = scene_->getCurrentJointPositions();
  q_target[0] = 0.5;
  q_target[1] = 0.5;
  const Eigen::Matrix4d fk_target = scene_->forwardKinematics(q_target, "link2");
  geometry_msgs::msg::Pose target = roboplan_ros_cpp::se3ToPose(fk_target);

  // Solve the IK using iMarker feedback
  visualization_msgs::msg::InteractiveMarkerFeedback feedback;
  feedback.event_type = visualization_msgs::msg::InteractiveMarkerFeedback::POSE_UPDATE;
  feedback.pose = target;
  const auto result = ik_->process_feedback(feedback);

  // Verify the result is non-trivial and close to the start
  ASSERT_TRUE(result.has_value());
  const Eigen::Matrix4d fk = scene_->forwardKinematics(*result, "link2");
  EXPECT_FALSE(result->isZero(1e-3));
  EXPECT_NEAR(fk(0, 3), target.position.x, 0.01);
  EXPECT_NEAR(fk(1, 3), target.position.y, 0.01);
  EXPECT_NEAR(fk(2, 3), target.position.z, 0.01);
}

}  // namespace roboplan_ros_visualization
